File: JavaScript/code/TPDict.js
import {calculateTilePatternProbabilities} from "./Helper.js"

export default class TPDict{
  constructor(input_samples, sizes, warp=null, borders=null){
    //rotation and flipping need to be added
    this._patterns = {};
    this._q_prob = {};
    if(!(input_samples[0][0] instanceof Array)){
      input_samples = [input_samples];
    }
    if(!(sizes instanceof Array)){
      sizes=[sizes];
    }
    sizes = sizes.slice();
    if (sizes.indexOf(1)<0){
      sizes.push(1);
    }
    const [probs, patterns, border_patterns] = calculateTilePatternProbabilities(input_samples, sizes, warp, borders);
    this._q_prob = probs;
    this._patterns = patterns;
    this._border_patterns = border_patterns;
  }

  getTPArray(size){
    return this._patterns[size];
  }

  getTPBorderArray(size, loc){
    return this._border_patterns[size][loc];
  }

  getQProbability(size){
    return this._q_prob[size];
  }
}

File: C#/code/TPDict.cs
using System;
using System.Collections.Generic;

namespace ETPKLDiv{
  class TPDict{
    private Dictionary<int, PatternCounts> _q_counts;
    private Dictionary<int, PatternList> _patterns;
    private Dictionary<int, BorderPatternList> _border_patterns;

    public TPDict(List<int[,]> input_samples, List<int> sizes, WarpOptions warp=null, BorderOptions borders=null){
      Tuple<Dictionary<int, PatternCounts>, Dictionary<int, PatternList>, Dictionary<int, BorderPatternList>> temp =
        Helper.CalculateTilePatternProbabilities(input_samples, sizes, warp, borders);
      this._q_counts = temp.Item1;
      this._patterns = temp.Item2;
      this._border_patterns = temp.Item3;
    }

    public PatternList GetTPArray(int size){
      return this._patterns[size];
    }

    public PatternList GetTPBorderArray(int size, string dir){
      return this._border_patterns[size][dir];
    }

    public PatternCounts GetQCount(int size){
      return this._q_counts[size];
    }
  }
}
